packages/labs/nextjs/src/index.ts

```diff
@@ -0,0 +1,85 @@
+
+// TODO(augustjk) Consider adding option for adding template shadowroot polyfill
+// automatically
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface LitSsrPluginOptions {}
+
+export = (_pluginOptions: LitSsrPluginOptions = {}) =>
+  (nextConfig: NextConfig = {}) => {
+    return Object.assign({}, nextConfig, {
+      webpack: (config, options) => {
+        const {isServer, nextRuntime, webpack} = options;
+        config.module.rules.unshift({
+          // Grab entry points for all pages
+          // TODO(augustjk) This may not work for the new "app" directory
+          // introduced in Next.js v13
+          test: /\/pages\//,
+          // Exclude Next's own distributed files as they're commonjs and won't
+          // play nice with `imports-loader`
+          exclude: /next\/dist\//,
+          loader: 'imports-loader',
+          options: {
+            imports: ['side-effects @lit-labs/ssr-react/enable-lit-ssr.js'],
+          },
+        });
+
+        if (isServer && nextRuntime === 'nodejs') {
+          // Next.js uses this externals setting to skip webpack bundling
+          // external code for server environments but we specifically want to
+          // catch react/jsx-runtime imports for external packages as well.
+          // This is what we're grabbing:
+          // https://github.com/vercel/next.js/blob/412dfc52cc5e378e4f74e44af698dad25031a938/packages/next/src/build/webpack-config.ts#L1378-L1429
+          const nextHandleExternals = config.externals[0];
+          config.externals = [
+            (opt: {request: string}) => {
+              if (
+                opt.request === 'react/jsx-dev-runtime' ||
+                opt.request === 'react/jsx-runtime'
+              ) {
+                // Returning empty promise makes these requests go through webpack
+                return Promise.resolve();
+              }
+              return nextHandleExternals(opt);
+            },
+          ];
+        }
+
+        config.plugins.push(
+          new webpack.NormalModuleReplacementPlugin(
+            /react/,
+            function (resource: {request: string; context: string}) {
+              if (
+                resource.request === 'react/jsx-runtime' &&
+                // Don't replace our own imports
+                !/labs\/ssr-react/.test(resource.context)
+              ) {
+                resource.request = '@lit-labs/ssr-react/jsx-runtime';
+              }
+              if (
+                resource.request === 'react/jsx-dev-runtime' &&
+                // Don't replace our own imports
+                !/labs\/ssr-react/.test(resource.context)
+              ) {
+                resource.request = '@lit-labs/ssr-react/jsx-dev-runtime';
+              }
+            }
+          )
+        );
+
+        // There's potential for conflict here if existing `webpack` config also
+        // overwrites externals
+        if (typeof nextConfig.webpack === 'function') {
```

Comment:
What is going on here? The user can pass their own webpack?

Reply:
Since we're wrapping/adding to the user provided next config, users could have a `webpack` as part of their own next config (which needs to be a function). This is the same pattern used by the Next team's own plugin here: https://github.com/vercel/next.js/blob/b2ac41c75493ab759bd8ecfd7ee111dc66c2db5c/packages/next-mdx/index.js#L38-L40

I should also clarify that better in the comment.

Reply:
Done.